lib/FrontendTool/ReferenceDependencies.cpp

```diff
@@ -23,253 +23,312 @@
-  std::sort(tmp.begin(), tmp.end(), [](const std::string &a,
-                                       const std::string &b) -> bool {
-              return std::lexicographical_compare(a.rbegin(), a.rend(),
-                                                  b.rbegin(), b.rend());
-            });
-  return tmp;
-}
+private:
+  void emitTopLevelNames(const ArrayRef<const Decl *> decls);
+  void emitTopLevelDecl(const Decl *D);
+  void emitExtensionDecl(const ExtensionDecl *ED);
+  void emitNominalTypeDecl(const NominalTypeDecl *NTD);
+  void emitValueDecl(const ValueDecl *VD) const;
 
-bool swift::emitReferenceDependencies(DiagnosticEngine &diags, SourceFile *SF,
-                                      DependencyTracker &depTracker,
-                                      StringRef outputPath) {
-  assert(SF && "Cannot emit reference dependencies without a SourceFile");
-  
+  void emitNominalTypes() const;
+  void emitMembers() const;
+  void emitDynamicLookupMembers(const SourceFile *SF) const;
+
+  void findNominalsAndOperators(const DeclRange members);
+  static bool extendedTypeIsPrivate(const TypeLoc inheritedType);
+  static bool declIsPrivate(const Decl *member);
+
+  template <typename Decl_T> static bool hasVisibleName(const Decl_T *decl) {
+    return decl->hasName() && isVisibleOutsideItsFile(decl);
+  }
+};
+
+class DependencyEmitter {
+  const SourceFile *SF;
+  const DependencyTracker &depTracker;
+  llvm::raw_ostream &out;
+
+public:
+  using MemberTableEntryTy = std::pair<ReferencedNameTracker::MemberPair, bool>;
+
+  DependencyEmitter(const SourceFile *SF, const DependencyTracker &depTracker,
+                    llvm::raw_ostream &out)
+      : SF(SF), depTracker(depTracker), out(out) {}
+
+  void emit() const;
+
+private:
+  void emitTopLevelNames(
+      ArrayRef<std::pair<DeclBaseName, bool>> sortedTopLevelNames) const;
+  void
+  emitMemberAccesses(const ArrayRef<MemberTableEntryTy> sortedMembers) const;
+  void emitNominalTypes(const ArrayRef<MemberTableEntryTy> sortedMembers) const;
+  void emitDynamicLookupMembers(const ArrayRef<std::pair<DeclBaseName, bool>>
+                                    sortedDynamicLookupNames) const;
+  void emitExternalReferences(const DependencyTracker &depTracker) const;
+};
+
+} // end anonymous namespace
+
+bool ReferenceDependencyFileEmitter::emit(DiagnosticEngine &diags,
+                                          SourceFile *SF,
+                                          const DependencyTracker &depTracker,
+                                          StringRef outputPath) {
   // Before writing to the dependencies file path, preserve any previous file
   // that may have been there. No error handling -- this is just a nicety, it
   // doesn't matter if it fails.
   llvm::sys::fs::rename(outputPath, outputPath + "~");
 
   std::error_code EC;
-  llvm::raw_fd_ostream out(outputPath, EC, llvm::sys::fs::F_None);
+  auto out = llvm::make_unique<llvm::raw_fd_ostream>(outputPath, EC,
```

Comment:
Why put this on the heap at all?

Reply:
I can't figure out how to return anything like an Optional raw_fd_ostream because it cannot be copied. If I try to std::move it, then I cannot create the None case for returning an error.

Reply:
Since the heap-allocation overhead occurs only once per primary file, it seems like the right tradeoff. However, if you have a better way, I'd love to learn!

Reply:
Hm, yeah, it looks like there's no move-constructor today. Okay.